### PR TITLE
Homework 4

### PR DESCRIPTION
**Pull request by yura182**

### CODE REVIEW

File: src/start.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import app from './app'
```
Comment (l2online92) [inline src/start.js]: It's a really good solution to split logic between two files: `app.js` and `start.js`!  But I would rename `start.js` into `server.js`

File: test/integration/endpoint.test.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+          return done(error)
+        }
+
+        expect(response.body).toEqual(expectedBody)
+        done()
+      })
+  }
+  )
+
+  function getUserData (login, password, age) {
```
Comment (l2online92) [inline test/integration/endpoint.test.js]: Let's move `getUserData` and `getExpectedUser` to the top of the file and outside of tests

File: test/unit/controller/userController.test.js [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+  })
+
+  test('should call user service with expected argument when retrieve user by id', () => {
+    userController.getUserById(request, response, next)
+
+    expect(userService.getUser).toHaveBeenCalledWith(ID)
+  })
+
+  test('should return expected HTTP status in response when user was not found by id', () => {
+    userService.getUser.mockReturnValue(undefined)
```
Comment (l2online92) [inline test/unit/controller/userController.test.js:136]: When user is not found it rather will be `null` than `undefined`

File: test/unit/controller/userController.test.js [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+
+    userController.getUserById(request, response, next)
+
+    expect(response.json).toHaveBeenCalledWith({ message: 'User not found' })
+  })
+
+  test('should call next callback with expected error when retrieve user by id failed', () => {
+    userService.getUser.mockImplementation(() => {
+      throw new Error(ERROR_MESSAGE)
```
Comment (l2online92) [inline test/unit/controller/userController.test.js]: I think you can also use `error` variable here: `throw error`, and check please other similar places in the code

File: test/unit/controller/userController.test.js (hunk 3 above)
Comment (yura182) [reply test/unit/controller/userController.test.js:136]: In current implementation user service returns `undefined` if user was not found
image

File: test/unit/controller/userController.test.js (hunk 4 above)
Comment (yura182) [reply test/unit/controller/userController.test.js]: Fixed

File: src/start.js (hunk 1 above)
Comment (yura182) [reply src/start.js]: Fixed

File: test/integration/endpoint.test.js (hunk 2 above)
Comment (yura182) [reply test/integration/endpoint.test.js]: Fixed. What is the purpose to move helper functions outside describe block to the top of the file?

**Pull request merged**